### PR TITLE
Extract services and runners

### PR DESCRIPTION
**Pull request by MartinHjelmare**
This mostly moves things around, besides the new setup functions that are called from the engine setup method. I've noted with inline comments where there are some other changes.

## Changes
- Add services package.
- Extract file server service.
- Extract terminal service.
- Add runners package.
- Move plugin.py to runners/subprocess.py.
- Extract subprocess runner.
- Allow registering more disconnect handlers.
- Allow registering more reset_engine handlers.
- Fix kill_all_plugins.

closes https://github.com/oeway/ImJoy/issues/180

### CODE REVIEW

File: imjoy/connection/handler.py [hunk 1]
```diff
@@ -420,313 +31,17 @@ def logging_callback(msg, type="info"):  # pylint: disable=redefined-builtin
         logger.debug("Client %s is not registered", sid)
         return {"success": False, "error": "client has not been registered"}
 
-    await kill_all_plugins(engine, sid)
+    # services and runners can register reset_engine_* handlers that will be called here
```
Comment (MartinHjelmare) [inline imjoy/connection/handler.py]: Note this.

File: imjoy/connection/handler.py [hunk 2]
```diff
@@ -762,53 +77,13 @@ def exited(_):
...
-            kill_plugin(engine, plugin["id"])
-
-
 @sio_on("disconnect", namespace=NAME_SPACE)
 async def disconnect(engine, sid):
     """Disconnect client."""
     logger = engine.logger
-    disconnect_client_session(engine, sid)
-    disconnect_plugin(engine, sid)
+    # services and runners can register disconnect_* handlers that will be called here
```
Comment (MartinHjelmare) [inline imjoy/connection/handler.py]: Note this.

File: imjoy/runners/subprocess.py [hunk 3]
```diff
@@ -175,20 +539,22 @@ def kill_plugin(engine, pid):
         del plugins[pid]
 
 
-async def kill_all_plugins(engine, ssid):
+@sio_on("reset_engine_plugins", namespace=NAME_SPACE)
+async def reset_engine(engine, sid, _):
```
Comment (MartinHjelmare) [inline imjoy/runners/subprocess.py]: We register a handler that will be called from the main `reset_engine` handler.

File: imjoy/runners/subprocess.py [hunk 4]
```diff
@@ -112,7 +475,8 @@ def add_plugin(engine, plugin_info, sid=None):
         plugin_info["sid"] = sid
 
 
-def disconnect_plugin(engine, sid):
+@sio_on("disconnect_plugin", namespace=NAME_SPACE)
+async def disconnect_plugin(engine, sid):
```
Comment (MartinHjelmare) [inline imjoy/runners/subprocess.py]: This will be called from the main `disconnect` handler.

File: imjoy/runners/subprocess.py [hunk 5]
```diff
@@ -68,7 +430,8 @@ def add_client_session(engine, session_id, client_id, sid, base_url, workspace):
     return client_connected
 
 
-def disconnect_client_session(engine, sid):
+@sio_on("disconnect_client_session", namespace=NAME_SPACE)
+async def disconnect_client_session(engine, sid):
```
Comment (MartinHjelmare) [inline imjoy/runners/subprocess.py]: This will be called from the main `disconnect` handler.

File: imjoy/runners/subprocess.py [hunk 6]
```diff
@@ -175,20 +539,22 @@ def kill_plugin(engine, pid):
...
-    tasks = []
-    for sid in list(plugin_sids.keys()):
+    for plugin_info in plugin_sids.values():
         try:
-            await on_kill_plugin(ssid, {"id": plugin_sids[sid]["id"]})
+            await on_kill_plugin(engine, sid, {"id": plugin_info["id"]})
         except Exception as exc:  # pylint: disable=broad-except
             logger.error("%s", exc)
 
-    return asyncio.gather(*tasks)
```
Comment (MartinHjelmare) [inline imjoy/runners/subprocess.py]: This wasn't used, so I removed it. We can enhance this later and use gather again and handle the exception etc.

File: imjoy/runners/subprocess.py (hunk 3 above)
Comment (oeway) [reply imjoy/runners/subprocess.py]: It seems other handler functions has the same name as the corresponding events, would it make more sense to use the same name here?

Comment (MartinHjelmare) [reply imjoy/runners/subprocess.py]: Yeah, we can use the same name. I change names on the event a couple of times and forgot to change the function name.

**Comment by oeway** [comment]
Maybe also make a package called `workers`? Perhaps also separate the code of Python 2 and 3.

**Comment by MartinHjelmare** [comment]
I can extract workers in a separate PR. I think after #59 Python 2 and 3 will be mostly separate. Did you mean something else?

**Comment by oeway** [comment]
Sounds good.

**Comment by MartinHjelmare** [comment]
Can we merge here?

**Comment by oeway** [comment]
Yes, I thought you are waiting for #59

**Comment by MartinHjelmare** [comment]
No, that is unrelated.

**Pull request merged**